src/probability/hitting_time.lean

```diff
@@ -0,0 +1,226 @@
+    push_neg at h_exists,
+    simp only [not_le.mpr hi, set.mem_Icc, false_iff, not_exists, and_imp],
+    exact λ k hkn hki, h_exists k ⟨hkn, hki.trans hi.le⟩, },
+end
+
+lemma hitting_lt_iff [is_well_order ι (<)] {m : ι} (i : ι) (hi : i ≤ m) :
+  hitting u s n m x < i ↔ ∃ j ∈ set.Ico n i, u j x ∈ s :=
+begin
+  split; intro h',
+  { have h : ∃ j ∈ set.Icc n m, u j x ∈ s,
+    { by_contra,
+      simp_rw [hitting, if_neg h, ← not_le] at h',
+      exact h' hi, },
+    exact ⟨hitting u s n m x, ⟨le_hitting_of_exists h, h'⟩, hitting_mem_set h⟩, },
+  { obtain ⟨k, hk₁, hk₂⟩ := h',
+    refine lt_of_le_of_lt _ hk₁.2,
+    exact hitting_le_of_mem hk₁.1 (hk₁.2.le.trans hi) hk₂, },
+end
+
+end inequalities
+
+/-- A discrete hitting time is a stopping time. -/
+lemma hitting_is_stopping_time
+  [conditionally_complete_linear_order ι] [is_well_order ι (<)] [encodable ι]
+  [topological_space β] [pseudo_metrizable_space β] [measurable_space β] [borel_space β]
+  {f : filtration ι m} {u : ι → α → β} {s : set β} {n n' : ι}
+  (hu : adapted f u) (hs : measurable_set s) :
+  is_stopping_time f (hitting u s n n') :=
+begin
+  intro i,
+  cases le_or_lt n' i with hi hi,
+  { have h_le : ∀ x, hitting u s n n' x ≤ i := λ x, (hitting_le x).trans hi,
+    simp [h_le], },
+  { have h_set_eq_Union : {x | hitting u s n n' x ≤ i} = ⋃ j ∈ set.Icc n i, u j ⁻¹' s,
+    { ext x,
+      rw [set.mem_set_of_eq, hitting_le_iff_of_lt _ hi],
+      simp only [set.mem_Icc, exists_prop, set.mem_Union, set.mem_preimage], },
+    rw h_set_eq_Union,
+    exact measurable_set.Union (λ j, measurable_set.Union_Prop $
+      λ hj, f.mono hj.2 _ ((hu j).measurable hs)) }
+end
+
+section complete_lattice
+
+variables [complete_lattice ι] {u : ι → α → β} {s : set β} {f : filtration ι m}
+
+lemma hitting_eq_Inf (x : α) : hitting u s ⊥ ⊤ x = Inf {i : ι | u i x ∈ s} :=
+begin
+  simp only [hitting, set.mem_Icc, bot_le, le_top, and_self, exists_true_left, set.Icc_bot,
+    set.Iic_top, set.univ_inter, ite_eq_left_iff, not_exists],
+  intro h_nmem_s,
+  symmetry,
+  rw Inf_eq_top,
+  exact λ i hi_mem_s, absurd hi_mem_s (h_nmem_s i),
+end
+
+end complete_lattice
+
+section nat
+
+variables {u : ℕ → α → β} {s : set β} {f : filtration ℕ m}
+
+lemma hitting_le_iff_nat {i n : ℕ} {x : α} (hx : ∃ j, j ≤ n ∧ u j x ∈ s) :
+  hitting u s 0 n x ≤ i ↔ ∃ j ≤ i, u j x ∈ s :=
+begin
+  cases lt_or_le i n with hi hi,
+  { rw hitting_le_iff_of_lt _ hi,
+    simp, },
+  { simp only [(hitting_le x).trans hi, true_iff],
+    obtain ⟨j, hj₁, hj₂⟩ := hx,
+    exact ⟨j, hj₁.trans hi, hj₂⟩, },
+end
+
+/-- A `ℕ`-indexed hitting time of an adpated process is a stopping time if the set it hits is
+measurable. -/
+lemma hitting_is_stopping_time_nat
+  [topological_space β] [pseudo_metrizable_space β] [measurable_space β] [borel_space β]
+  {f : filtration ℕ m} {u : ℕ → α → β} (hu : adapted f u) (hs : measurable_set s) (n : ℕ) :
+  is_stopping_time f (hitting u s 0 n) :=
+hitting_is_stopping_time hu hs
```

Comment:
That lemma can be removed, since `hitting_is_stopping_time` can be used to prove this.